### DISCUSSION TITLE
PCA9685 with ESP32 and wifi performance - Watchdog timeout

### DISCUSSION THREAD

**Discussion by sjgcornish** [ArminJo/ServoEasing#73] [General]
servoEasing is a very nice and clever bit of software however when using PCA9685 and controlling the servos over WIFI there is an unpredictable lag between command and action. I am not entirely sure how servoEasing works using interrupts for servo sync, my thoughts are that he PCA has its own 50Hz clock and it sends its stored commands to the servos on a 20ms loop, this means that you only have to send a command via I2C when needed to make a change in position such as when receiving a command over WIFI.  I assume that servoEasing has its own 20ms interrupt which syncs the servos and it is also updating all the the servos every 20ms, which is required if you are not using the PCA9685.  For some reason  it appears that, WIFI, servoEasing with PCA9685 functionality do not play well together, perhaps it is related to yield() not working in the 'while' loop. One solution for me is to use a dedicated non WIFI Arduino for servo control and just send it commands from the ESP32 over serial but then you need call back to the ESP32 when the move is completed (I have implemented this mechanism using the ESPservo library)

Or could it be due to ..
// The ESP32 I2C interferes with the Ticker / Timer library used.
// Even with 100 kHz clock we have some dropouts / NAK's because of sending address again instead of first data.

**Comment by ArminJo** [comment]
Please let a Led change its state if you received a command from wifi and issuing a command for servo easing to determine if wifi lags.
Do you use blocking wait functions for servo movements? Maybe enabling line 1616 and 1628 in ServoEasing.hpp helps in this case.
Please report.

**Reply by sjgcornish** [reply]
Made the two changes as suggested. This came up moving from 170 degrees to 25 degrees slowly (20) This is repeatable i.e. it doe sit every time

Move to degrees: 170 at speed: 20 degrees per second
Move to degrees: 25 at speed: 20 degrees per second
E (199640) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (199640) task_wdt:  - async_tcp (CPU 1)
E (199640) task_wdt: Tasks currently running:
E (199640) task_wdt: CPU 0: IDLE0
E (199640) task_wdt: CPU 1: IDLE1
E (199640) task_wdt: Aborting.
abort() was called at PC 0x401430f8 on core 0

This code is being called in response to a command sent over WIFI.  Note the dots represent the mapping degr to values for each servo. I believe this is non blocking code. The ESP32 devkit v4 does not have a user LED on board and I have not got around to adding one yet.

void setDirectionMap(int degr, int speed){
    Serial.print("Move to degrees: "); .....
    ......
	ServoEasing::ServoEasingNextPositionArray[0] = BLDegrees;
    ServoEasing::ServoEasingNextPositionArray[1] = FLDegrees;
    setEaseToForAllServosSynchronizeAndStartInterrupt(speed);

    while (ServoEasing::areInterruptsActive()) {
        delay(1);
        yield();
    }
}

I understand that this means that async_tcp  is getting starved of execution time, then after a timeout period, the watchdog gets triggered and resets the chip. My code has
#include <AsyncTCP.h>
#include <ESPAsyncWebServer.h> 
I note from a google search that the ESP rebooting is a fairly common problem for various reason. I have also used the Adafruit PCA9685 library and not had any problems. I suspect that the during easing the the easing task if a long one is not yielding resulting in something! that causes that results in the reboot. I have crudely simulated easing in to Loop using previousmillieconds -currentmilliseconds with yield at the end of the loop an this works Ok even using a   an HTML joystick over a TCP socket
 
I am not sure if this helps but the stack trace is..

0x40088fb8: invoke_abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/panic.c line 156
0x40089235: abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/panic.c line 171
0x401430f8: task_wdt_isr at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/task_wdt.c line 174
0x4015594f: esp_pm_impl_waiti at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/pm_esp32.c line 492
0x401449f3: esp_vApplicationIdleHook at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/freertos_hooks.c line 63
0x4008ba05: prvIdleTask at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/freertos/tasks.c line 3382
0x4008a246: vPortTaskWrapper at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/freertos/port.c line 143

**Reply by ArminJo** [reply]
I have no clue, but you can try inserting a yield after the line 3413 in ServoEasing.hpp
```
for (uint_fast8_t tServoIndex = 0; tServoIndex <= ServoEasing::sServoArrayMaxIndex; ++tServoIndex) {
yield();
```

**Reply by ArminJo** [reply]
Typo! It is 2413. Function updateAllServos()

**Reply by sjgcornish** [reply]
Getting closer to the problem at least for the unplanned resets. I am using webserial.h to send move commands and the error that chrome dev tools reports is ..

 'Unchecked runtime.lastError: The message port closed before a response was received'

If I programmatically move the servo from 25 degrees to 170 degrees slowly then there is no reset but it is OK if I use webserial to move at speed 60.  This ties in with the Asynch_TCP error reported by Arduino serial monitor and must be due to servoEasing code not yielding to a task associated with AsyncTCP which during a servo movement so the watchdog timer event is triggered because it is stating there waiting to respond to the tcp request.  The only reason why the yield()is not working is because yield() nor delay() work during a  callback or when used in an ISR. I must admit to having difficulty following your code (I am only a hobby programmer) so I so not know if what part of your code is in and ISR and part of a callback.

I suspect that the slow response is again due to the asyncTCP but any delay is not enough to cause the watchdog to trigger and hence a reset.

Note yield() only yields to higher priority tasks and the watchdog timer is a low priority task. delay() yields to all tasks including the watch dog timer that triggers the watchdog event.

**Reply by ArminJo** [reply]
> Note yield() only yields to higher priority tasks and the watchdog timer is a low priority task. delay() yields to all tasks including the watch dog timer that triggers the watchdog event.

Thanks for this information 👍 
Reset can eventually avoided by inserting  `esp_task_wdt_reset();` instead of (or additional to?)  `yield` at line 2413.

**Reply by sjgcornish** [reply]
That did not work, however I think I have solved the problem, more testing required but the problem was partially mine of the making. My WebSerial.msgCallback(recvMsg) was taking to long and the delay combined with the delay in processing slow long servo movements tripped the watch dog time and it appears for the sometime slow response. I modified the recvMsg callback and put all the code in the main loop. 
This whole process has been an education for me.

**Comment by sjgcornish** [comment]
This makes interesting reading

https://www.reddit.com/r/esp32/comments/w2q3r7/esp32_unexplained_slowness_with_simple_iteration/
